### PR TITLE
First round of optimizations for vector functions.

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR merges the `vectors-optimize-brute-force` feature branch, which makes the following changes to how vector functions are computed:
* #45390
* #45936
* #46103
* #46155
* #46190

These changes give the following improvements on two test datasets:

**random-s-100-angular (cosine similarity)**

```
Algorithm                          Recall      QPS
EsBruteforce()                     1.000       23.205
EsBruteforce(optimized)            1.000       63.789
```

**random-s-100-euclidean (L2 norm)**

```
Algorithm                          Recall      QPS
EsBruteforce()                     1.000       38.852
EsBruteforce(optimized)            1.000       64.361
```

This round of optimizations focused on low-hanging fruit and does not make any API changes. We are tracking ideas for more complex improvements in #46202.

Co-authored by @mayya-sharipova.

### CODE REVIEW

File: x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java [hunk 1]
```diff
@@ -180,8 +182,11 @@ public void parse(ParseContext context) throws IOException {
 
         // encode array of floats as array of integers and store into buf
         // this code is here and not int the VectorEncoderDecoder so not to create extra arrays
-        byte[] buf = new byte[dims * INT_BYTES];
-        int offset = 0;
+        byte[] bytes = indexCreatedVersion.onOrAfter(Version.V_7_4_0) ? new byte[dims * INT_BYTES + INT_BYTES] : new byte[dims * INT_BYTES];
```
Comment (mayya-sharipova) [inline x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java]: should we change this to 7_5_0 version?

**Comment by mayya-sharipova** [comment]
@jtibshirani Thanks, great improvement!

File: x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java (hunk 1 above)
Comment (jtibshirani) [reply x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapper.java]: Thanks, I had forgotten!

File: x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapperTests.java [hunk 2]
```diff
@@ -89,6 +107,53 @@ public void testDefaults() throws Exception {
         );
     }
 
+    public void testAddDocumentsToIndexBefore_V_7_4_0() throws Exception {
+        Version indexVersion = Version.V_7_3_0;
```
Comment (mayya-sharipova) [inline x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapperTests.java]: looks like we also need to update the version in tests

Comment (jtibshirani) [reply x-pack/plugin/vectors/src/test/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorFieldMapperTests.java]: Thanks again, I've been having a sleepy day.

**Review by mayya-sharipova — APPROVED** [review]
@jtibshirani Thanks, excellent optimizations

**Comment by jtibshirani** [comment]
A note for future readers: it seems QPS has dropped ~10% moving from ES version 7.6 to 7.7: 

**random-s-100-angular (cosine similarity)**

```
Algorithm             Recall   QPS
EsBruteforce(7.6)     1.000    66.989
EsBruteforce(7.7)     1.000    60.906
```

I suspect this is caused by the fact we started [compressing binary doc values](https://issues.apache.org/jira/browse/LUCENE-9211) in Lucene 8.5.

**Pull request merged**